---
title: "Data Pre-Processing"
author: "Lucas Gillenwater"
date: "`r Sys.Date()`"
output: html_document
---


# Necessary Libraries

Prior to running this script you may need to first download the necessary libraries. 

```{r}
source("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/R/data_prep_functions.R")
library(Publish)
library(knitr)
library(datawizard)
library(tidyverse)
```
# Data Files

This script assumes that you are running it from the parent directory. The data subdirectory is organized as:
- data
  - HTP_data
  - processed
  - pathway

Also, it assumes that you have the necessary data files. 

```{r}
# Reading in the data
cytokine_file <- "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/HTP_data/MSD_200413_EXPLORER_noMeta_Exp3456_firstPIDsampleOnly_meansOnly_FINAL_JRS_v5.tsv"
metabolite_file <- "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/HTP_data/P4C_LCMS_abundance_long_TrisomExplorer_08092020.tsv"
metadata_file <-"~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/HTP_data/P4C_metadata_021921_Costello.txt"
comorbidity_file <- "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/HTP_data/P4C_Comorbidity_020921.tsv"
transcriptomic_file <- "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/HTP_data/HTP_WholeBlood_RNAseq_FPKMs_Synapse.txt"
```

# Processing Data into a single object 

``` {r}
htp = read_HTP(cytokine_file, metabolite_file, metadata_file, comorbidity_file)

```

# Filter to only data with matched cytokine and metabolite profiles

``` {r}
htp_overlap <- get_overlap(cytokine_data = htp$cytokines,
						   metabolite_data = htp$metabolites, 
						   clinic_data = htp$clinic)
```

# Create 'Table 1' of cohort demographics
```{r}

colnames(htp_overlap$clinic) = gsub(" |/", "_", names(htp_overlap$clinic))
tab1 <- univariateTable(Karyotype~ Sex+Age_at_visit + BMI + Sample_source   +Anxiety+Any_autoimmune_skin_condition+Any_congenital_heart_defect+Any_hearing_loss_condition+ Any_hypothyroidism+Any_seizure_history+Any_sleep_apnea+                Asthma_reactive_airway_disease +Celiac_disease+Depression+ Frequent_Recurrent_pneumonia+Obesity+Recurrent_otitis_media  , htp_overlap$clinic)
kable(summary(tab1))
## write.csv(summary(tab1), file = "./results/tables/Table1.csv", row.names = F) ## For initial writing, but afterwards this is updated in format for the publication. No need to update again. 

```
# Log Transform the data
``` {r}
## Log transform
htp_overlap[1:2] = lapply(htp_overlap[1:2], log2)


```
# PCA plots of the data

``` {r}

cyt_pca <- prcomp(htp_overlap[[1]])$x
ggplot(as.data.frame(cyt_pca), aes(x = PC1, y = PC2)) +
        geom_text(aes(label = rownames(cyt_pca)))


met_pca <- prcomp(htp_overlap[[2]])$x
ggplot(as.data.frame(met_pca), aes(x = PC1, y = PC2)) +
        geom_text(aes(label = rownames(met_pca)))

```

# Remove Outliers
The PCA plot of the cytokine data indicates several samples that fell outside the expected range for cytokine concentrations. Therefore, these individuals were tagged as outliers and removed from the analysis. 

``` {r}

# Remove outlier subjects
toRemove = c("HTP0553A", "HTP0565A", "HTP0572A","HTP0589A","HTP0664A", "HTP0555B")
htp_overlap[1:2] = remove_subj(htp_overlap[1:2], toRemove)
htp_overlap$clinic = htp_overlap$clinic[rownames(htp_overlap[[1]]),]


```

# Adjusting for confounding variables
Several demographic variables are known to confound the relationships between molecular feature and co-occurring condition. Moreover, this data was collected at different time points ('sample_source'). Therefore, these data were linearly adjusted using a mixed effects model for each of these confounding variables. 'Age' and 'sex' were input as fixed effects. The 'sample_source' was input as a random effect. This was done for consistency with other studies by the Human Trisome Project

``` {r}
full_data = lapply(list(htp_overlap[[1]],htp_overlap[[2]]), function(x) merge(x,htp_overlap$clinic, by.x = 0, by.y = "LabID"))
names(full_data[[1]]) =  make.names(names(full_data[[1]]))
names(full_data[[2]]) =  make.names(names(full_data[[2]]))

to_exclude <- make.names(names(htp_overlap$clinic))

mixed = lapply(1:2, function(x) adjust(full_data[[x]], effect = "Sample_source",
                                       exclude = to_exclude ,
                                       multilevel = T, keep_intercept = TRUE))

mixed = lapply(1:2, function(x) adjust(mixed[[x]], effect = "Age_at_visit", exclude = to_exclude, keep_intercept = TRUE))
mixed = lapply(1:2, function(x) adjust(mixed[[x]], effect = "Sex", exclude = to_exclude, keep_intercept = TRUE))

mixed = lapply(1:2, function(x) mixed[[x]][make.names(names(htp_overlap[[x]]))])
names(mixed[[1]]) = names(htp_overlap[[1]])
names(mixed[[2]]) = names(htp_overlap[[2]])
mixed = lapply(mixed, function(x) as.data.frame(t(x)))

names(mixed[[1]]) = rownames(htp_overlap[[1]])
names(mixed[[2]]) = rownames(htp_overlap[[1]])

```
# Separating T21 and D21 data
``` {r}
T21.mixed = only_T21(list(c(mixed, list(htp_overlap$clinic))))
D21.mixed = only_D21(list(c(mixed, list(htp_overlap$clinic))))
all.mixed = list(cytokines = mixed[[1]], metabolites = mixed[[2]], clinic =  htp_overlap$clinic)

```

# Gene Expression Data
```{r}
#### This chunk  preps the HTP transcriptomic data for downstream analysis
## Processing workflow:
##    - load count data
##    - log transform
##    - apply variance filter
##    - pivot to wider

## load expression data
expression <- read.delim(transcriptomic_file)

## find the number of subjects
n <- length(unique(expression$LabID))

## filter for those with less than 20% missingness
missingness <- expression %>%
    group_by(EnsemblID) %>%
    summarise(missingness = sum(Value ==0)/n) 

miss_filtered <- expression %>%
    filter(EnsemblID %in% (missingness %>% filter(missingness < .2) %>% .$EnsemblID))

## logtranformation
miss_filtered$logValue <- log(miss_filtered$Value+1)


## filtering based on variance threshold
var_filtered <- miss_filtered %>%
    group_by(EnsemblID) %>%
    summarise(variance = var(logValue)) %>%
    filter(variance > .01)
    

variance_filtered <- miss_filtered %>%
    filter(EnsemblID %in% var_filtered$EnsemblID) %>%
    as_tibble() 


### protein_filtered
prot_filtered <- variance_filtered %>%
    filter(Gene_type == "protein_coding")

## pivot wider. Have to include the unique identifier for row numbers to avoid any errors. 
htp_expr <- prot_filtered %>%
    dplyr::select(LabID, Gene_name,Value) %>%
    distinct(LabID, Gene_name, .keep_all = T) %>%
    pivot_wider(names_from = Gene_name, id_cols = LabID, values_from = Value)



### get the mappings of gene_name, ensembl, and chr from expression
gene_map <- expression %>%
    dplyr::select(Gene_name, EnsemblID,Chr) %>%
    filter(EnsemblID %in% variance_filtered$EnsemblID) %>%
    as_tibble()



## Adjust the transcriptomic data for age, sex, and sample source as the cytokine and metabolite data
metadata <- htp_overlap$clinic

expr_metadata <- htp_expr %>%
    inner_join(metadata, by = "LabID")

original_names <- names(htp_expr)
names(expr_metadata) <- make.names(names(expr_metadata))

to_exclude <- make.names(names(metadata))

mixed <- adjust(expr_metadata, effect = "Sample_source",
                exclude = to_exclude ,
                multilevel = T, keep_intercept = TRUE)

mixed <- adjust(mixed, effect = "Age_at_visit", exclude = to_exclude, keep_intercept = TRUE)
mixed <- adjust(mixed, effect = "Sex", exclude = to_exclude, keep_intercept = TRUE)

names(mixed) <- names(expr_metadata)
mixed <- mixed %>%
    column_to_rownames("LabID")

mixed <- mixed %>% dplyr::select(!(make.names(names(metadata)[-5])))

htp_expr <- mixed

expression_list<- list(expression = htp_expr,gene_map = gene_map)
```

### Filter all the datasets to only include the overlapping set

``` r
T21.mixed <- lapply(T21.mixed, function(x) x %>% filter(rownames(x) %in% rownames(expression_list$expression)))
D21.mixed <- lapply(D21.mixed, function(x) x %>% filter(rownames(x) %in% rownames(expression_list$expression)))

```

### save the datasets

``` r
save(T21.mixed, file = "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/processed/T21_mixed_cyt_met.Rdata")
save(D21.mixed, file = "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/processed/D21_mixed_cyt_met.Rdata")
save(expression_list, file = "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/processed/processed_transcriptome.RData")

```


```{r}
sessionInfo()
```

