# Comparing RNA clustering in T21 with integrated clustering overall

# heatmap
``` r
load(file ="~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/clusterings/rna_clust.RData")
load(file ="~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/clusterings/all_clust.RData")
load(file ="~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/clusterings/cyt_clust.Rdata")
load(file ="~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/clusterings/met_clust.RData")
load("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/processed/processed_transcriptome.RData")
##load("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/processed/processed_transcriptome_standardized.RData")
load("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/processed/T21_standardized_cyt_met.RData")
source("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/R/cluster_evaluation.R")
source("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/R/data_prep_functions.R")

T21.transcripts <- expression_list$expression %>%
    filter(rownames(.) %in% rownames(T21.mixed$cytokines))

rna_differential_expression <- diff_expr_wilcoxon(t(T21.transcripts),rna_clust$clustering)

rna_cluster_associations <- association_test(clinic = T21.mixed$clinic,
                                         clustering = rna_clust$clustering,
                                         phenotypes = names(T21.mixed$clinic)[10:21])

p1 <- clusterHeatmap(omics.data = t(T21.transcripts),
               diff_expr = rna_differential_expression,
               clustering = rna_clust$clustering,
               cluster_assoc = rna_cluster_associations,
               threshold = .01,
               split_num = 1,
               title = "T21 Transcripts",
               fontsize = 0,
               phenotypes = names(T21.mixed$clinic)[10:21],
               clinical_threshold = .2,
               column_names = F,
               dendrogram = F
               )

pdf("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/cluster_heatmaps/RNA_diff_clinic_D4clusters_25neighbors_heatmap.pdf", width = 12, height = 8)

draw(p1$h, heatmap_legend_side = "bottom")

dev.off()

```

```r
library(ggalluvial)

cyt_met_all_clusterings <- cbind(cytokine = cyt_clust$clustering,
                                 integrated = all_clust$clustering,
                                 metabolite = met_clust$clustering,
                                 rna = rna_clust$clustering)
cyt_met_all_clusterings <- cyt_met_all_clusterings %>%
    as.data.frame() %>%
    count(cytokine, integrated,metabolite, rna)

cyt_met_all_clusterings[1:4] <- apply(cyt_met_all_clusterings[1:4],2,as.factor)

cols <- c("blue4", "orange4", "springgreen4", "brown3", "grey")
names(cols) <- c(1,2,3,4,"D21")

p3 <- ggplot(as.data.frame(cyt_met_all_clusterings),
       aes(y = n, axis1 = integrated, axis2 = rna)) +
  geom_alluvium(aes(fill = integrated), width = 1/12) +
  geom_stratum(width = 1/12, fill = "black", color = "grey") +
  geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
    scale_x_discrete(limits = c("integrated","rna" ), expand = c(.05, .05)) +
    scale_fill_manual(values = cols) +
#  scale_fill_brewer(type = "qual", palette = "Set1") +
    ggtitle("") +
    theme_classic() +
    ylab("count")+
    theme(axis.text = element_text(size = 12),
          legend.position = "bottom")

ggsave(p3,file = "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/riverplots/rna_integrated_25neighbors_riverplot.pdf", height = 3, width = 5)


```

## permutations
```r
library(aricode) ## for AMI

set.seed(1234) ## for reproducibility


### calculating jaccard based on counting pairs between two clusterings
all_clustering <- all_clust$clustering

rna_clustering <- rna_clust$clustering

clust_num <- as.numeric(table(rna_clustering)) ## for checking the breakdown
total <- sum(clust_num) ## total size of the cohort


clusteringPairsJaccard <- function(integrated_clustering, other_clustering) {
    ss <- 0
    sd <- 0
    ds <- 0
    for(i in 1:4) {
        all_pairs <- expand.grid(names(integrated_clustering[integrated_clustering==i]),names(integrated_clustering[integrated_clustering==i]))
        all_pairs <- all_pairs[all_pairs$Var1 != all_pairs$Var2,]
        all_pairs$pair <- paste(all_pairs$Var1,all_pairs$Var2)
        ds_tmp <- nrow(all_pairs)    
        for(j in 1:4) {
            other_pairs <- expand.grid(names(other_clustering[other_clustering==j]),names(other_clustering[other_clustering==j]))
            other_pairs <- other_pairs[other_pairs$Var1 != other_pairs$Var2,]
            other_pairs$pair <- paste(other_pairs$Var1,other_pairs$Var2)
            ss_tmp <- length(intersect(all_pairs$pair, other_pairs$pair))
            sd_tmp <- nrow(other_pairs) - ss_tmp
            ss <- ss + ss_tmp
            sd <- sd + sd_tmp
            ds_tmp <- ds_tmp - ss_tmp
        }
        ds <- ds+ds_tmp
    }
    jaccard <- ss/(ss+sd+ds)
    return(jaccard)
}


rna_jaccard <- clusteringPairsJaccard(integrated_clustering = all_clustering,
                                      other_clustering = rna_clustering)

cluster_names <- names(rna_clustering)

permutation_jaccards <- sapply(1:1000, function(x) {
    permutation <- sample(rna_clust$clustering, size = total, replace = F)
    names(permutation) <- cluster_names
    clusteringPairsJaccard(integrated_clustering = all_clustering,
                                              other_clustering = permutation)
} )


empirical_p <- sum(rna_jaccard< permutation_jaccards)/1001


toplot <- as.data.frame(permutation_jaccards)

p4 <- ggplot(toplot, aes(x = permutation_jaccards)) +
    geom_density() +
    geom_vline(xintercept = rna_jaccard, color = "darkred") +
    theme_classic()

ggsave(p4, file = "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/clustering_permutations/permutation_density.pdf", width = 6, height = 3)


```


## RNA signatures for IMSs

``` r
library(fgsea)
pathways <- gmtPathways("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/pathway/h.all.v2023.1.Hs.symbols.gmt")


differential_expression <- diff_expr_limma(t(T21.transcripts),all_clust$clustering)







signature_output <- differential_expression$FC %>%
    full_join(differential_expression$sig, by ="feature") %>%
    arrange(q.value.1)

temp <- cbind(T21.transcripts, all_clust$clustering)
names(temp)[ncol(temp)] <- "clust"

ggplot(temp, aes(x = ACKR3)) + geom_density()
ggplot(temp, aes(x = ACKR3, fill = as.factor(clust))) + geom_density(alpha = .5)

signature_output <- signature_output[, c("feature", "FC.1", "q.value.1", "FC.2", "q.value.2", "FC.3", "q.value.3", "FC.4", "q.value.4")]

sig_1 <- signature_output %>%
    filter(q.value.1 < .15 & q.value.2 > .15 & q.value.3 > .15 & q.value.4 > .15)


plot(signature_output$FC.1,-log10(signature_output$q.value.1))


write.csv(signature_output, file = "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/tables/RNA_IMS_signatures.csv")

any(sig_list[[1]] %in% sig_list[[3]])
 

signatures <- lapply(1:4, function(i) {
                  rnk <- as.numeric(differential_expression$FC[,c(i+1)])
                  names(rnk) <- differential_expression$FC$feature
                  fgsea(pathways = pathways,
                  stats    = rnk,
                  minSize  = 15,
                  maxSize  = 500)
})


signatures_NES <- lapply(1:4, function (x) {    
    signatures[[x]] <- signatures[[x]] %>%
        select(pathway, NES) %>%
        setNames(c("pathway", paste0("NES",x)))
})


signatures_df <- signatures_NES %>% reduce(full_join, by = "pathway") %>% column_to_rownames("pathway")


signatures_padj <- lapply(1:4, function (x) {    
    signatures[[x]] <- signatures[[x]] %>%
        select(pathway, padj) %>%
        setNames(c("pathway", paste0("padj",x)))
})


signatures_df_padj <- signatures_padj %>% reduce(full_join, by = "pathway") %>% column_to_rownames("pathway")

signatures_df_padj %>% filter(

load(file = "~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/mediation_heatmaps/global_pathway_heatmaps/T21_heatmap.Rdata")

toplot <- t(signatures_df)

colnames(toplot) <- signatures_df$pathway
colnames(toplot) <- gsub("HALLMARK_|_", " ", colnames(toplot))

path_order <- row_order(p1)

col_fun <- colorRamp2(c(0.5,1,1.5), c("purple","white", "orange"))
toplot <- abs(toplot)
enrichemnt_heatmap <- Heatmap(t(toplot),
                              row_order= path_order,
                              show_row_dend = F,
                              col = col_fun,
                              column_names_side = "top",
                              column_order = rownames(toplot),
                              show_column_dend = F
                              )
pdf("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/results/IMS_RNA_signatures/individual_IMS_enrichment_signatures_GSEA.pdf")
draw(enrichemnt_heatmap)
dev.off()

```




### Sanity check

``` r
counts <- read.delim("./data/HTP_data/HTP_WholeBlood_RNAseq_Counts_Synapse.txt")

htp_expr <- counts %>%
    filter(Gene_name %in% names(T21.transcripts)) %>%
    filter(LabID %in% rownames(T21.transcripts)) %>%
    dplyr::select(LabID, Gene_name,Value) %>%
    distinct(LabID, Gene_name, .keep_all = T) %>%
    pivot_wider(names_from = Gene_name, id_cols = LabID, values_from = Value) %>%
    column_to_rownames("LabID")

htp_expr <- log(htp_expr +1e-10)

counts_limma <- diff_expr_limma(t(htp_expr), all_clust$clustering)

plot(counts_limma$'1.logFC', -log10(counts_limma$'1.adj.P.Val'))

counts_wilcox <- diff_expr_wilcoxon(t(htp_expr), all_clust$clustering)

plot(-counts_wilcox[,2], -log10(counts_wilcox[,3]))

cor.test(counts_limma$'1.logFC', counts_wilcox[,2])

cor.test(counts_limma$'2.adj.P.Val', counts_wilcox$p.value.x)

plot(counts_limma$'2.adj.P.Val', counts_wilcox$sig$q.value.1)

plot(counts_limma$'1.logFC', counts_wilcox$FC$FC.1)

FPKM <- read.delim("./data/HTP_data/HTP_WholeBlood_RNAseq_FPKMs_Synapse.txt")

htp_expr_FPKM <- FPKM %>%
    filter(Gene_name %in% names(T21.transcripts)) %>%
    filter(LabID %in% rownames(T21.transcripts)) %>%
    dplyr::select(LabID, Gene_name,Value) %>%
    distinct(LabID, Gene_name, .keep_all = T) %>%
    pivot_wider(names_from = Gene_name, id_cols = LabID, values_from = Value) %>%
    column_to_rownames("LabID") %>%
    log()

FPKM_limma <- diff_expr_limma(t(htp_expr_FPKM), all_clust$clustering, T21.mixed$clinic)

cor.test(counts_limma$'1.logFC', FPKM_limma$'1.logFC')

load("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/processed/processed_transcriptome.RData")
load("~/OneDrive - The University of Colorado Denver/Projects/ImmunoMetabolicSubtypes/data/processed/processed_transcriptome_standardized.RData")

adjusted <- expression_list$expression %>%
    filter(rownames(.) %in% rownames(T21.mixed$cytokines))

adjusted_limma <- diff_expr_limma(t(adjusted),all_clust$clustering)

adjusted_limma <- adjusted_limma %>%
    filter(feature %in% counts_limma$feature)

cor.test(counts_limma$'1.logFC', adjusted_limma$'1.logFC')
cor.test(FPKM_limma$'1.logFC', adjusted_limma$'1.logFC')

adjusted_wilcox <- diff_expr_wilcoxon(t(adjusted),all_clust$clustering)

adjusted_wilcox$FC <- adjusted_wilcox$FC %>%
    filter(feature %in% counts_limma$feature)

cor.test(adjusted_limma$'1.logFC', adjusted_wilcox$FC$FC.1)

cor.test(adjusted_limma$'1.logFC', adjusted_wilcox$FC$FC.1)

```
